Can't compile contract with Base58Decode interop in global scope

neo-go version 0.94.0-pre-55-g492a89eb

I have contract with `binary.Base58Decode` call in global scope.

```go
package contract
import (
	...
)
var (
	ctx storage.Context
	k1 = binary.Base58Decode([]byte("..."))
)
```

And I can't compile it. It worked fine in 0.93.0 though.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x88 pc=0xc6d2d1]

goroutine 1 [running]:
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).inlineCall(0xc0000ccb40, 0xc0000ab340, 0xc000391280)
        github.com/nspcc-dev/neo-go/pkg/compiler/inline.go:24 +0x111
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).Visit(0xc0000ccb40, 0x10dfa80, 0xc000391280, 0x0, 0x0)
        github.com/nspcc-dev/neo-go/pkg/compiler/codegen.go:849 +0x3799
go/ast.Walk(0x10d6240, 0xc0000ccb40, 0x10dfa80, 0xc000391280)
        go/ast/walk.go:52 +0x63
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).Visit(0xc0000ccb40, 0x10dff40, 0xc000391300, 0x0, 0x0)
        github.com/nspcc-dev/neo-go/pkg/compiler/codegen.go:512 +0x2170
go/ast.Walk(0x10d6240, 0xc0000ccb40, 0x10dff40, 0xc000391300)
        go/ast/walk.go:52 +0x63
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).convertGlobals.func1(0x10dff40, 0xc000391300, 0x10e0001)
        github.com/nspcc-dev/neo-go/pkg/compiler/codegen.go:308 +0x9c
go/ast.inspector.Visit(0xc000387f50, 0x10dff40, 0xc000391300, 0x10d79a0, 0xc000387f50)
        go/ast/walk.go:373 +0x3a
go/ast.Walk(0x10d79a0, 0xc000387f50, 0x10dff40, 0xc000391300)
        go/ast/walk.go:52 +0x63
go/ast.walkDeclList(0x10d79a0, 0xc000387f50, 0xc0003ba500, 0xc, 0x10)
        go/ast/walk.go:38 +0x9e
go/ast.Walk(0x10d79a0, 0xc000387f50, 0x10dfe00, 0xc0003b8700)
        go/ast/walk.go:353 +0x2346
go/ast.Inspect(...)
        go/ast/walk.go:385
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).convertGlobals(0xc0000ccb40, 0xc0003b8700, 0xc0003cf450)
        github.com/nspcc-dev/neo-go/pkg/compiler/codegen.go:303 +0x7d
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).traverseGlobals.func2(0xc0003e40b0)
        github.com/nspcc-dev/neo-go/pkg/compiler/analysis.go:96 +0xbd
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).ForEachPackage(0xc0000ccb40, 0xc00029e3b0)
        github.com/nspcc-dev/neo-go/pkg/compiler/compiler.go:71 +0x44
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).traverseGlobals(0xc0000ccb40, 0xc00029e4b8, 0xc0000ccb40, 0x140)
        github.com/nspcc-dev/neo-go/pkg/compiler/analysis.go:92 +0x1fd
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).compile(0xc0000ccb40, 0xc0003025c0, 0xc0003e40b0, 0xc0003e40b0, 0xc000386810)
        github.com/nspcc-dev/neo-go/pkg/compiler/codegen.go:1953 +0xe5
github.com/nspcc-dev/neo-go/pkg/compiler.CodeGen(0xc0003025c0, 0xb, 0x0, 0x0, 0xc0003025c0, 0x0, 0x0)
        github.com/nspcc-dev/neo-go/pkg/compiler/codegen.go:2020 +0x91
github.com/nspcc-dev/neo-go/pkg/compiler.CompileWithDebugInfo(0x7ffdee846f2e, 0xb, 0x0, 0x0, 0x1, 0xc0003a8458, 0x4, 0x1, 0x0, 0xc00029e6a8)
        github.com/nspcc-dev/neo-go/pkg/compiler/compiler.go:160 +0xaa
github.com/nspcc-dev/neo-go/pkg/compiler.CompileAndSave(0x7ffdee846f2e, 0xb, 0xc00029ea68, 0x6, 0x0, 0x0, 0x0, 0x0)
        github.com/nspcc-dev/neo-go/pkg/compiler/compiler.go:176 +0x226
github.com/nspcc-dev/neo-go/cli/smartcontract.contractCompile(0xc0003c6420, 0x10100, 0xc0003c6420)
        github.com/nspcc-dev/neo-go/cli/smartcontract/smart_contract.go:455 +0x449
github.com/urfave/cli.HandleAction(0xd49cc0, 0xffd788, 0xc0003c6420, 0xc000392700, 0x0)
        github.com/urfave/cli@v1.20.0/app.go:490 +0x82
github.com/urfave/cli.Command.Run(0xe98194, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, 0xeb9f2d, 0x27, 0x0, ...)
        github.com/urfave/cli@v1.20.0/command.go:210 +0x9fb
github.com/urfave/cli.(*App).RunAsSubcommand(0xc00039e680, 0xc0003c6160, 0x0, 0x0)
        github.com/urfave/cli@v1.20.0/app.go:379 +0x8d4
github.com/urfave/cli.Command.startApp(0xe99497, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, 0xebaf75, 0x28, 0x0, ...)
        github.com/urfave/cli@v1.20.0/command.go:298 +0x85f
github.com/urfave/cli.Command.Run(0xe99497, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, 0xebaf75, 0x28, 0x0, ...)
        github.com/urfave/cli@v1.20.0/command.go:98 +0x11e8
github.com/urfave/cli.(*App).Run(0xc00039e4e0, 0xc000030090, 0x9, 0x9, 0x0, 0x0)
        github.com/urfave/cli@v1.20.0/app.go:255 +0x768
main.main()
        command-line-arguments/main.go:18 +0x4e
```

--- a/pkg/compiler/analysis.go
+++ b/pkg/compiler/analysis.go
@@ -46,13 +46,24 @@ func (c *codegen) traverseGlobals() (int, int, int) {
 	var n, nConst int
 	initLocals := -1
 	deployLocals := -1
-	c.ForEachFile(func(f *ast.File, _ *types.Package) {
+	c.ForEachFile(func(f *ast.File, pkg *types.Package) {
 		nv, nc := countGlobals(f)
 		n += nv
 		nConst += nc
 		if initLocals == -1 || deployLocals == -1 || !hasDefer {
 			ast.Inspect(f, func(node ast.Node) bool {
 				switch n := node.(type) {
+				case *ast.GenDecl:
+					if n.Tok == token.VAR {
+						for i := range n.Specs {
+							for _, v := range n.Specs[i].(*ast.ValueSpec).Values {
+								num := c.countLocalsCall(v, pkg)
+								if num > initLocals {
+									initLocals = num
+								}
+							}
+						}
+					}
 				case *ast.FuncDecl:
 					if isInitFunc(n) {
 						num, _ := c.countLocals(n)

--- a/pkg/compiler/func_scope.go
+++ b/pkg/compiler/func_scope.go
@@ -102,6 +102,50 @@ func (c *funcScope) analyzeVoidCalls(node ast.Node) bool {
 	return true
 }
 
+func (c *codegen) countLocalsCall(n ast.Expr, pkg *types.Package) int {
+	ce, ok := n.(*ast.CallExpr)
+	if !ok {
+		return -1
+	}
+
+	var size int
+	var name string
+	switch fun := ce.Fun.(type) {
+	case *ast.Ident:
+		var pkgName string
+		if pkg != nil {
+			pkgName = pkg.Path()
+		}
+		name = c.getIdentName(pkgName, fun.Name)
+	case *ast.SelectorExpr:
+		name, _ = c.getFuncNameFromSelector(fun)
+	default:
+		return 0
+	}
+	if inner, ok := c.funcs[name]; ok && canInline(name) {
+		sig, ok := c.typeOf(ce.Fun).(*types.Signature)
+		if !ok {
+			info := c.buildInfo.program.Package(pkg.Path())
+			sig = info.Types[ce.Fun].Type.(*types.Signature)
+		}
+		for i := range ce.Args {
+			switch ce.Args[i].(type) {
+			case *ast.Ident:
+			case *ast.BasicLit:
+			default:
+				size++
+			}
+		}
+		// Variadic with direct var args.
+		if sig.Variadic() && !ce.Ellipsis.IsValid() {
+			size++
+		}
+		innerSz, _ := c.countLocalsInline(inner.decl, inner.pkg, inner)
+		size += innerSz
+	}
+	return size
+}
+
 func (c *codegen) countLocals(decl *ast.FuncDecl) (int, bool) {
 	return c.countLocalsInline(decl, nil, nil)
 }
@@ -117,40 +161,7 @@ func (c *codegen) countLocalsInline(decl *ast.FuncDecl, pkg *types.Package, f *f
 	ast.Inspect(decl, func(n ast.Node) bool {
 		switch n := n.(type) {
 		case *ast.CallExpr:
-			var name string
-			switch fun := n.Fun.(type) {
-			case *ast.Ident:
-				var pkgName string
-				if pkg != nil {
-					pkgName = pkg.Path()
-				}
-				name = c.getIdentName(pkgName, fun.Name)
-			case *ast.SelectorExpr:
-				name, _ = c.getFuncNameFromSelector(fun)
-			default:
-				return false
-			}
-			if inner, ok := c.funcs[name]; ok && canInline(name) {
-				sig, ok := c.typeOf(n.Fun).(*types.Signature)
-				if !ok {
-					info := c.buildInfo.program.Package(pkg.Path())
-					sig = info.Types[n.Fun].Type.(*types.Signature)
-				}
-				for i := range n.Args {
-					switch n.Args[i].(type) {
-					case *ast.Ident:
-					case *ast.BasicLit:
-					default:
-						size++
-					}
-				}
-				// Variadic with direct var args.
-				if sig.Variadic() && !n.Ellipsis.IsValid() {
-					size++
-				}
-				innerSz, _ := c.countLocalsInline(inner.decl, inner.pkg, inner)
-				size += innerSz
-			}
+			size += c.countLocalsCall(n, pkg)
 			return false
 		case *ast.FuncType:
 			num := n.Results.NumFields()

--- a/pkg/compiler/inline.go
+++ b/pkg/compiler/inline.go
@@ -19,6 +19,14 @@ func (c *codegen) inlineCall(f *funcScope, n *ast.CallExpr) {
 	pkg := c.buildInfo.program.Package(f.pkg.Path())
 	sig := c.typeOf(n.Fun).(*types.Signature)
 
+	// When inlined call is used during global initialization
+	// there is no func scope, thus this if.
+	if c.scope == nil {
+		c.scope = &funcScope{}
+		c.scope.vars.newScope()
+		defer func() { c.scope = nil }()
+	}
+
 	// Arguments need to be walked with the current scope,
 	// while stored in the new.
 	oldScope := c.scope.vars.locals

--- a/pkg/compiler/inline_test.go
+++ b/pkg/compiler/inline_test.go
@@ -115,6 +115,32 @@ func TestInline(t *testing.T) {
 	})
 }
 
+func TestInlineGlobalVariable(t *testing.T) {
+	t.Run("simple", func(t *testing.T) {
+		src := `package foo
+		import "github.com/nspcc-dev/neo-go/pkg/compiler/testdata/inline"
+		var a = inline.Sum(1, 2)
+		func Main() int {
+			return a
+		}`
+		eval(t, src, big.NewInt(3))
+	})
+	t.Run("complex", func(t *testing.T) {
+		src := `package foo
+		import "github.com/nspcc-dev/neo-go/pkg/compiler/testdata/inline"
+		var a = inline.Sum(3, 4)
+		var b = inline.SumSquared(1, 2) 
+		var c = a + b
+		func init() {
+			c--
+		}
+		func Main() int {
+			return c
+		}`
+		eval(t, src, big.NewInt(15))
+	})
+}
+
 func TestInlineConversion(t *testing.T) {
 	src1 := `package foo
 	import "github.com/nspcc-dev/neo-go/pkg/compiler/testdata/inline"

--- a/pkg/compiler/syscall_test.go
+++ b/pkg/compiler/syscall_test.go
@@ -4,10 +4,12 @@ import (
 	"math/big"
 	"testing"
 
+	"github.com/nspcc-dev/neo-go/pkg/core/interop/interopnames"
 	istorage "github.com/nspcc-dev/neo-go/pkg/core/interop/storage"
 	"github.com/nspcc-dev/neo-go/pkg/interop/contract"
 	"github.com/nspcc-dev/neo-go/pkg/interop/storage"
 	"github.com/nspcc-dev/neo-go/pkg/smartcontract/callflag"
+	"github.com/nspcc-dev/neo-go/pkg/vm"
 	"github.com/nspcc-dev/neo-go/pkg/vm/stackitem"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
@@ -71,4 +73,22 @@ func TestNotify(t *testing.T) {
 	assert.Equal(t, exp0, s.events[0].Item.Value())
 	assert.Equal(t, "single", s.events[1].Name)
 	assert.Equal(t, []stackitem.Item{}, s.events[1].Item.Value())
+}
+
+func TestSyscallInGlobalInit(t *testing.T) {
+	src := `package foo
+		import "github.com/nspcc-dev/neo-go/pkg/interop/binary"
+		var a = binary.Base58Decode([]byte("5T"))
+		func Main() []byte {
+			return a
+		}`
+	v, s := vmAndCompileInterop(t, src)
+	s.interops[interopnames.ToID([]byte(interopnames.SystemBinaryBase58Decode))] = func(v *vm.VM) error {
+		s := v.Estack().Pop().Value().([]byte)
+		require.Equal(t, "5T", string(s))
+		v.Estack().PushVal([]byte{1, 2})
+		return nil
+	}
+	require.NoError(t, v.Run())
+	require.Equal(t, []byte{1, 2}, v.Estack().Pop().Value())
 }

--- a/pkg/interop/binary/binary.go
+++ b/pkg/interop/binary/binary.go
@@ -42,7 +42,7 @@ func Base58Encode(b []byte) string {
 // Base58Decode decodes given base58 string represented as a byte slice into
 // a new byte slice. It uses `System.Binary.Base58Decode` syscall.
 func Base58Decode(b []byte) []byte {
-	return neogointernal.Syscall1("System.Binary.Base64Decode", b).([]byte)
+	return neogointernal.Syscall1("System.Binary.Base58Decode", b).([]byte)
 }
 
 // Itoa converts num in a given base to string. Base should be either 10 or 16.

